django-stubs/test/client.pyi

```diff
@@ -151,7 +151,10 @@ class _RequestFactory(Generic[_T]):
     ) -> _T: ...
 
 class RequestFactory(_RequestFactory[WSGIRequest]): ...
-class _AsyncRequestFactory(_RequestFactory[_T]): ...
+
+class _AsyncRequestFactory(_RequestFactory[_T]):
+    defaults: dict[str, Any]
```

Comment:
Nice find with `asgiref.typing.HTTPScope`! I think that's what we _want_ here. But I also see that e.g. `client` is typed as `client: Optional[Tuple[str, int]]` but Django does: `"client": ["127.0.0.1", 0],` 🙂

Not sure what to make of it.. I have nothing against adopting `HTTPScope` anyways though.

I can't think of any free-form properties, but I would suggest starting out strict and relax the type later on to `dict[str, Any]` if anyone runs into troubles.

I wouldn't say we need to add `asgiref`? As you say Django already depend on it, it should follow along. Don't have any strong opinions about it either

Reply:
> `client: Optional[Tuple[str, int]]` but Django does: `"client": ["127.0.0.1", 0]`

Interesting. Since you can't typehint a list as `list[str, int]`, it is probably a decent compromise though.

> I wouldn't say we need to add asgiref?

My preference would be to add it to deps and not rely on transitive dependencies. Although granted, the failure mode with missing dependencies in type stubs is very minor.

Reply:
According to asgiref docs https://asgi.readthedocs.io/en/latest/specs/www.html#http-connection-scope `client` is `Iterable[str, int]`. So yeah, a compromise.

Reply:
I've adjusted to `defaults: asgiref.typing.HTTPScope` and added `asgiref` as a first line dependency in `setup.py`

Reply:
Also, I moved the `defaults` declaration to each class existing runtime: `RequestFactory` and `AsyncRequestFactory` to please the inheritance. I think that's most accurate? Let me know if you see any adjustments

Reply:
> Also, I moved the defaults declaration to each class existing runtime: RequestFactory and AsyncRequestFactory to please the inheritance. I think that's most accurate? Let me know if you see any adjustments

I'm not sure. I think the previous approach was slightly closer to Django's implementation, even if Django violates Liskiv substitution principle here (as it does in lots of places). But these are internal protocols, so doesn't really matter as long as all classes using them are updated as needed.

You would need to add the `defaults` attribute to all classes that derive from `_RequestFactory` hierarchy, it seems you missed `Client` and `AsyncClient`.

Reply:
Yeah, I looked again and was about to say just that moving it it instead becomes misplaced. I'll add an ignore instead.